// parser-combinator/build/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.T = exports.X = exports.MD = exports.JSON = exports.N = exports.C = exports.F = exports.parser = exports.stream = exports.standard = exports.parsec = exports.genlex = exports.data = undefined;

var _index = require('./data/index');

var _index2 = _interopRequireDefault(_index);

var _index3 = require('./genlex/index');

var _index4 = _interopRequireDefault(_index3);

var _index5 = require('./parsec/index');

var _index6 = _interopRequireDefault(_index5);

var _parser = require('./parsec/parser');

var _parser2 = _interopRequireDefault(_parser);

var _index7 = require('./standard/index');

var _index8 = _interopRequireDefault(_index7);

var _index9 = require('./stream/index');

var _index10 = _interopRequireDefault(_index9);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var JSON = _index8.default.jsonParser; /*
                                        * Parsec
                                        * https://github.com/d-plaindoux/parsec
                                        *
                                        * Copyright (c) 2016 Didier Plaindoux
                                        * Licensed under the LGPL2 license.
                                        */

var MD = _index8.default.markdownBundle;
var X = _index8.default.extractorBundle;
var T = _index8.default.tokenBundle;

exports.data = _index2.default;
exports.genlex = _index4.default;
exports.parsec = _index6.default;
exports.standard = _index8.default;
exports.stream = _index10.default;
exports.parser = _parser2.default;
exports.F = _index5.F;
exports.C = _index5.C;
exports.N = _index5.N;
exports.JSON = JSON;
exports.MD = MD;
exports.X = X;
exports.T = T;
