// parser-combinator/build/test/standard/markdown/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _titleTest = require('./title-test');

var _titleTest2 = _interopRequireDefault(_titleTest);

var _textParserTest = require('./text-parser-test');

var _textParserTest2 = _interopRequireDefault(_textParserTest);

var _bulletParserTest = require('./bullet-parser-test');

var _bulletParserTest2 = _interopRequireDefault(_bulletParserTest);

var _codeLineParserTest = require('./code-line-parser-test');

var _codeLineParserTest2 = _interopRequireDefault(_codeLineParserTest);

var _singleLinesParserTest = require('./single-lines-parser-test');

var _singleLinesParserTest2 = _interopRequireDefault(_singleLinesParserTest);

var _documentParserTest = require('./document-parser-test');

var _documentParserTest2 = _interopRequireDefault(_documentParserTest);

var _fileParserTest = require('./file-parser-test');

var _fileParserTest2 = _interopRequireDefault(_fileParserTest);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    titleTest: _titleTest2.default,
    textTest: _textParserTest2.default,
    bulletParser: _bulletParserTest2.default,
    codeLineParser: _codeLineParserTest2.default,
    singleLinesParser: _singleLinesParserTest2.default,
    documentParserTest: _documentParserTest2.default,
    fileParserTest: _fileParserTest2.default
};
